access-widener-javaparser/src/main/java/net/fabricmc/accesswidener/javaparser/SourceAccessWidenerTransformer.java

```diff
@@ -0,0 +1,178 @@
+import com.github.javaparser.ast.Modifier;
+import com.github.javaparser.ast.NodeList;
+import com.github.javaparser.ast.body.ConstructorDeclaration;
+import com.github.javaparser.ast.body.FieldDeclaration;
+import com.github.javaparser.ast.body.MethodDeclaration;
+import com.github.javaparser.ast.body.TypeDeclaration;
+import com.github.javaparser.ast.body.VariableDeclarator;
+import com.github.javaparser.ast.nodeTypes.NodeWithName;
+import com.github.javaparser.resolution.UnsolvedSymbolException;
+
+import net.fabricmc.accesswidener.AccessWidener;
+import net.fabricmc.accesswidener.EntryTriple;
+
+public class SourceAccessWidenerTransformer {
+	protected final AccessWidener widener;
+
+	public SourceAccessWidenerTransformer(AccessWidener widener) {
+		this.widener = widener;
+
+		if (!widener.isSourceCompatible()) {
+			throw new UnsupportedOperationException("AccessWidener must have source compatible entries!");
+		}
+	}
+
+	/**
+	 * A heuristic on whether the java file may be transformed.
+	 */
+	public boolean mayTransform(String javaFilePath) {
+		String pkg = AccessWidener.getPackage(javaFilePath);
+		return this.widener.getPackages().contains(pkg);
+	}
+
+	public boolean transform(CompilationUnit unit) {
+		boolean transformed = false;
+
+		for (TypeDeclaration<?> type : unit.findAll(TypeDeclaration.class)) {
+			String qualifiedName = getFullyQualifiedName(type).orElse(null);
+
+			if (qualifiedName == null) {
+				System.err.println("Unable to find internal name of local class in " + unit.getPrimaryType()
+						.flatMap(SourceAccessWidenerTransformer::getFullyQualifiedName) + ".java");
+				return false;
+			}
+
+			if (this.widener.getTargets().contains(qualifiedName)) {
+				boolean isInterface = type.isClassOrInterfaceDeclaration() && type.asClassOrInterfaceDeclaration()
+						.isInterface() || type.isAnnotationDeclaration();
+
+				String internalName = qualifiedName.replace('.', '/');
+				NodeList<Modifier> classModifiers = type.getModifiers();
+				int classFlags = SourceUtil.toClassFlags(classModifiers, type);
+				int newFlags = this.widener.getClassAccess(internalName).apply(classFlags, internalName, classFlags);
+				SourceUtil.fromFlags(classModifiers, newFlags, false);
+
+				// todo add unsealing support when javaparser supports it
+
+				for (MethodDeclaration declaration : type.getMethods()) {
+					try {
+						String desc = declaration.toDescriptor();
+						EntryTriple triple = EntryTriple.create(internalName, declaration.getNameAsString(), desc, true);
+						this.apply(declaration.getModifiers(), triple, AccessWidener::getMethodAccess, classFlags, isInterface);
+					} catch (UnsolvedSymbolException ignore) {
+						// unresolved name, just don't apply the AW, maybe a warn?
+						// if we do a warning, ihmo it should check if a simple name matches (or if imports match or if it's qualified then the full name)
+						// it will be an expensive check, however it's an error case so it should be fine
```

Comment:
Silently ignoring something isn't great.. 

I don't think we really need the symbol solver for this, inferring the type or methods+fields+subclass handling isn't really using it?

Resolving methods may require explicit support for specialized overloads (return type may be a subtype of the AW declared return type) and generic arguments/rval. Source transformation is also more tricky in normally requiring to propagate widening down across the hierarchy while the jvm spec cares less about it being consistent.

Reply:
infering type does require the symbolsolver, it is technically possible to match without it I think, but it's difficult